### PR TITLE
[SPARK-12317][SQL]Support units (m,k,g) in SQLConf

### PR DESCRIPTION
**Pull request by kevinyu98**
Hello: adding the configure value for  AUTO_BROADCASTJOIN_THRESHOLD and DEFAULT_SIZE_IN_BYTES.

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/SQLConf.scala [hunk 1]
```diff
@@ -93,7 +94,7 @@ private[spark] object SQLConf {
           isPublic: Boolean = true): SQLConfEntry[Int] =
       SQLConfEntry(key, defaultValue, { v =>
         try {
-          v.toInt
+          Utils.byteStringAsBytes(v).toInt
```
Comment (srowen) [inline sql/core/src/main/scala/org/apache/spark/sql/SQLConf.scala]: It looks like you're making all integer config values parse as if they're memory sizes. That's not right, right?

Comment (kevinyu98) [inline sql/core/src/main/scala/org/apache/spark/sql/SQLConf.scala]: Hello Sean: Thanks for your comment. Yes, you are right. There are other methods are not meaning to use memory sizes. (like COLUMN_BATCH_SIZE, etc).  There are couple approaches, can  you suggest which way is preferable way for this problem or suggest a new way to fix this ?
1. we will document that [g|G|m|M|k|K] means memory size.

2. create a new method of intConf for AUTO_BROADCASTJOIN_THRESHOLD.

3.  create a rule to the parseByteString. like K/KB/M/MB means 1024, k/kb/m/mb means 1000.

Thanks.

**Comment by watermen** [comment]
@kevinyu98 I think we can add a new method like `intConfWithMem`, and also need to check the value in bytes need to LessThanOrEqual `Int.MaxValue`.

`K/KB/M/MB` is equals to `k/kb/m/mb`, right? See the code from JavaUtils below.
```
  private static long parseByteString(String str, ByteUnit unit) {
    String lower = str.toLowerCase().trim();
```

**Comment by kevinyu98** [comment]
@watermen thanks for your input. It is good idea if we decide to go with approach 2. create a new method of intConf for AUTO_BROADCASTJOIN_THRESHOLD.  
If we decide to go with approach 3, then we may need to change the parseByteString part to distinct lower and upper case. 

@srowen what do you think? Thanks.

**Comment by srowen** [comment]
It does seem like you would want a separate utility method for handling configuration values that are intended to be memory sizes.

Don't make up new behavior; just use the behavior from the `Utils` method. In particular, technically, "k" means 1000 and "Ki" means 1024, but I think we follow the JVM memory string conventions instead.

I don't see why `toString.toInt` solves an overflow problem. You get a `long` back, and if one particular value must be an `int`, simply handle argument checking for that value directly. No need for new methods.

**Comment by kevinyu98** [comment]
@srowen Thanks Sean, I will create a new method of intConf for AUTO_BROADCASTJOIN_THRESHOLD. Will update the PR soon.

**Comment by srowen** [comment]
@kevinyu98  are you still working on this one?

**Comment by kevinyu98** [comment]
@srowen Hello Sean, yes, sorry for the delay. I will submit the updated PR today.

**Comment by kevinyu98** [comment]
@srowen Hello Sean: I am sorry that I have't been able to submit the PR yet. I will conti. work on it tomorrow.

**Comment by kevinyu98** [comment]
@srowen Hello Sean: I have submit the new code, can you help review? Thanks a lot.

File: sql/core/src/main/scala/org/apache/spark/sql/SQLConf.scala [hunk 2]
```diff
@@ -107,7 +135,7 @@ private[spark] object SQLConf {
         isPublic: Boolean = true): SQLConfEntry[Long] =
       SQLConfEntry(key, defaultValue, { v =>
         try {
-          v.toLong
+          Utils.byteStringAsBytes(v)
```
Comment (srowen) [inline sql/core/src/main/scala/org/apache/spark/sql/SQLConf.scala]: This is still wrong

File: sql/core/src/main/scala/org/apache/spark/sql/SQLConf.scala [hunk 3]
```diff
@@ -100,6 +101,33 @@ private[spark] object SQLConf {
         }
       }, _.toString, doc, isPublic)
 
+    def intMemConf(
```
Comment (srowen) [inline sql/core/src/main/scala/org/apache/spark/sql/SQLConf.scala]: The spacing and alignment are wrong throughout this method

File: sql/core/src/main/scala/org/apache/spark/sql/SQLConf.scala [hunk 4]
```diff
@@ -100,6 +101,33 @@ private[spark] object SQLConf {
...
+                    key: String,
+                    defaultValue: Option[Int] = None,
+                    doc: String = "",
+                    isPublic: Boolean = true): SQLConfEntry[Int] =
+      SQLConfEntry(key, defaultValue, { v =>
+        var isNegative: Boolean = false
```
Comment (srowen) [inline sql/core/src/main/scala/org/apache/spark/sql/SQLConf.scala]: Why `var`? assign the result of the try to a `val`. No need for a type.

File: sql/core/src/main/scala/org/apache/spark/sql/SQLConf.scala [hunk 5]
```diff
@@ -100,6 +101,33 @@ private[spark] object SQLConf {
...
+        try {
+          isNegative = (v.toInt < 0)
+        } catch {
+          case _: Throwable =>
```
Comment (srowen) [inline sql/core/src/main/scala/org/apache/spark/sql/SQLConf.scala]: Why `Throwable`? I assume you mean `NumberFormatException`

File: sql/core/src/main/scala/org/apache/spark/sql/SQLConf.scala [hunk 6]
```diff
@@ -100,6 +101,33 @@ private[spark] object SQLConf {
...
+        }
+        if (!isNegative) {
```
Comment (srowen) [inline sql/core/src/main/scala/org/apache/spark/sql/SQLConf.scala]: It's clearer to flip this conditional.

File: sql/core/src/main/scala/org/apache/spark/sql/SQLConf.scala [hunk 7]
```diff
@@ -100,6 +101,33 @@ private[spark] object SQLConf {
...
+          if ((Utils.byteStringAsBytes(v) <= Int.MaxValue.toLong) &&
```
Comment (srowen) [inline sql/core/src/main/scala/org/apache/spark/sql/SQLConf.scala]: Why compute this 3 times?

**Comment by srowen** [comment]
Are there other configs that should take a memory string?

File: sql/core/src/main/scala/org/apache/spark/sql/SQLConf.scala (hunk 3 above)
Comment (kevinyu98) [inline sql/core/src/main/scala/org/apache/spark/sql/SQLConf.scala]: I will make the changes. I used the code format and run the scalastyle test, thought it passed the styles. I will look more carefully next time. Thanks.

File: sql/core/src/main/scala/org/apache/spark/sql/SQLConf.scala (hunk 4 above)
Comment (kevinyu98) [inline sql/core/src/main/scala/org/apache/spark/sql/SQLConf.scala]: ok, I will make the change.

File: sql/core/src/main/scala/org/apache/spark/sql/SQLConf.scala (hunk 5 above)
Comment (kevinyu98) [inline sql/core/src/main/scala/org/apache/spark/sql/SQLConf.scala]: yah, I want to catch the exception, then do nothing. I will make the changes.

File: sql/core/src/main/scala/org/apache/spark/sql/SQLConf.scala (hunk 6 above)
Comment (kevinyu98) [inline sql/core/src/main/scala/org/apache/spark/sql/SQLConf.scala]: ok

File: sql/core/src/main/scala/org/apache/spark/sql/SQLConf.scala (hunk 7 above)
Comment (kevinyu98) [inline sql/core/src/main/scala/org/apache/spark/sql/SQLConf.scala]: I will put it in a variable. thanks.

File: sql/core/src/main/scala/org/apache/spark/sql/SQLConf.scala (hunk 2 above)
Comment (kevinyu98) [inline sql/core/src/main/scala/org/apache/spark/sql/SQLConf.scala]: sorry, I thought it was only one place. But actually there are two places. I will create a new method for SHUFFLE_TARGET_POSTSHUFFLE_INPUT_SIZE. Thanks.

**Comment by srowen** [comment]
@kevinyu98 are you still working on this?

**Comment by kevinyu98** [comment]
Hello Sean: Sorry for the delay. Yes, I have made most code changes, and I will try to finish it up soon and do more testing. Will keep you updated. Thanks, Happy New Year !

**Comment by kevinyu98** [comment]
@srowen Hello Sean: Sorry for taking so long. Can you review the code? Thanks. Kevin

File: sql/core/src/main/scala/org/apache/spark/sql/SQLConf.scala [hunk 8]
```diff
@@ -100,6 +101,37 @@ private[spark] object SQLConf {
         }
       }, _.toString, doc, isPublic)
 
+    def intMemConf(
+        key: String,
+        defaultValue: Option[Int] = None,
+        doc: String = "",
+        isPublic: Boolean = true): SQLConfEntry[Int] =
+      SQLConfEntry(key, defaultValue, { v =>
+        val isInteger: Option[Int] = try {
```
Comment (srowen) [inline sql/core/src/main/scala/org/apache/spark/sql/SQLConf.scala]: I find this body simpler:

```
try {
  v.toInt
} catch {
  case _: NumberFormatException =>
    val sizeInBytes = try {
      Utils.byteStringAsBytes(v)
    } catch {
      case _: NumberFormatException =>
        throw new IllegalArgumentException(s"$key should be int, but was $v")
    }
    require(getSizeAsBytes <= Int.MaxValue && getSizeAsBytes >= Int.MinValue)
    sizeInBytes.toInt
}
```

I'd like to hear from @marmbrus or @viirya for example before merging a change like this.

**Comment by marmbrus** [comment]
This seems like a nice convenience for the users.  A few comments:
 - The description / title could be improved.  Some think about support units for configuration.
 - Do we really need both an Int and a Long version? (maybe we do)
 - Sean's suggestions seems good, but I would include a better error than the default for the `require`.

**Comment by kevinyu98** [comment]
Hello @srowen @marmbrus @viirya : I have made the code changes, and change the title based on the comments. Can you help review the codes? Thanks.

**Comment by viirya** [comment]
I have the same question as @marmbrus, do we need two methods for int and long? Looks like they are very similar.

**Comment by viirya** [comment]
Besides, -1 and -1g will be the same or different behavior?

**Comment by kevinyu98** [comment]
Hello @viirya : Good Point. I just test it, -1 and -1g will have different behavior. It will take -1, and throw IllegalArgumentException for -1g. Thanks.

**Comment by kevinyu98** [comment]
@viirya @yhuai @srowen @marmbrus @concretevitamin: SHUFFLE_TARGET_POSTSHUFFLE_INPUT_SIZE was added by spark-9850, PR#9276, I am including Yin , 
Michael set the autoBroadcastJoinThreshold to 10 * 1024 * 1024 through PR3064,
spark-2393 set the autoBroadcastJoinThreshold to Int.
I am not sure which one to choose either, so I cc on the persons who introduce these two fields. Thanks for your input.

File: sql/core/src/main/scala/org/apache/spark/sql/SQLConf.scala [hunk 9]
```diff
@@ -214,7 +256,7 @@ private[spark] object SQLConf {
       doc = "When true, enable partition pruning for in-memory columnar tables.",
       isPublic = false)
 
-  val AUTO_BROADCASTJOIN_THRESHOLD = intConf("spark.sql.autoBroadcastJoinThreshold",
+  val AUTO_BROADCASTJOIN_THRESHOLD = intMemConf("spark.sql.autoBroadcastJoinThreshold",
```
Comment (yhuai) [inline sql/core/src/main/scala/org/apache/spark/sql/SQLConf.scala]: How about we just use `longMemConf`?

**Comment by marmbrus** [comment]
I think just `long` is fine for now.  The title still has spelling errors and is really long.  I would say:

`[SPARK-12317][SQL] Support units (m,k,g) in SQLConf`. In the description of the PR (which becomes the rest of the commit message) you can spell out which options you change and more details.

**Comment by kevinyu98** [comment]
Hello @marmbrus, thanks. So your mean I can remove the code change for intMemConf, and keep the code for longMemConf for this jira? I will make the PR title and description changes. I need to close this PR, open another one, seems there is some issues when I did last git push.

**Pull request closed without merging**